### PR TITLE
Implements wavelet denoising

### PR DESCRIPTION
**Pull request by stsievert**
This pull request implements wavelet denoising (as mentioned in issue #602). This pull request uses [PyWavelets] to implement this.

- [x] add tests to cover 100% of code (tested each optional argument)
- [x] add example to gallery (added to plot_denoise.py, currently untested b/c couldn't compile)
- [ ] add reference to `CONTRIBUTORS.txt` (final step to avoid merge conflicts)
- [x] Only allow specifying noise variance. There's a paper by Donoho et al that specifies more detail.
- [x] Only compute wavelet transform once on image. I believe this requires renaming the helper function to something like `_soft_threshold` with an optional parameter for the threshold. If this parameter is None, we can estimate the threshold from the noise variance, similar to #1837 
- [x] Test! Also, reformat the tests for removal of optional argument

@grlee77 also had useful points
- [x] generalize to n dimensional wavelet transforms (not doing it on each color plane)
- [x] automatically choose threshold. This is currently implemented, but there is an open PR to estimate the noise std in an image, PR #1837.
- [x] allow choice between soft and hard thresholding

Running the example in docstring yields the following results:

figure_1

### Code changes
* added `denoise_wavelet` in `_denoise.py`
* added PyWavelets to `optional_requirements.txt`.

### Unsure about
* clipping. Wavelet denoising may result in values that are beyond the max/min of normal images. A keyword argument for this has been added that clips between 0/1 for the skimage workflow -- is that the right approach?
* sizing. There are cases where `pywt.threshold` returns an array of different size than the input (e.g., in my testing, for odd sized arrays when using `db2`). Currently, the code enforces that the output should be the same size as the image -- is that the right approach?
* documentation formatting. I tried building the docs (typing `make` in `docs/`) but couldn't find any function definitions.
* tests. How should tests be implemented for this?

[PyWavelets]:http://pywavelets.readthedocs.org/en/latest/

### CODE REVIEW

**Comment by stefanv** [comment]
Thanks for this contribution, Scott.  We are in the middle of releasing 0.12, so please be patient for feedback, and ping us again if you haven't heard back within a week or so after the release.  Thanks!

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -355,3 +355,92 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+        coeffs = pywt.wavedec2(img, wavelet)
+        denoised_coeffs = [pywt.threshold(c, value=threshold) for c in coeffs]
+        return pywt.waverec2(denoised_coeffs, wavelet)
+
+    size_error = ValueError("The image is not of the correct shape. Try "
+                            "passing in a different sized image (i.e the "
+                            "'db2' wavelets only accept even image sizes")
+
+    im_type = im.dtype
+    if not im_type.kind == 'f':
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: why do you need extra variable here?

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -355,3 +355,92 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+    """
+    def _denoise_wavelet(img, wavelet, threshold):
+        import pywt
+        coeffs = pywt.wavedec2(img, wavelet)
+        denoised_coeffs = [pywt.threshold(c, value=threshold) for c in coeffs]
+        return pywt.waverec2(denoised_coeffs, wavelet)
+
+    size_error = ValueError("The image is not of the correct shape. Try "
+                            "passing in a different sized image (i.e the "
+                            "'db2' wavelets only accept even image sizes")
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: I suggest to give user an option to choose the way this case is handled, not force him to preprocess data

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -355,3 +355,92 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+                                                           threshold=threshold)
+            except ValueError:
+                raise size_error
+
+    else:
+        raise ValueError('only 2-d and 3-d images may be denoised with this '
+                         'function')
+
+    if out.shape != im.shape:
+        raise size_error
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: It is better to handle all the unsupported cases at the beggining of a function.
Also, use should prepare a test for each of them.

File: skimage/restoration/_denoise.py [hunk 4]
```diff
@@ -355,3 +355,92 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+        out = _denoise_wavelet(img, wavelet=wavelet, threshold=threshold)
+
+    elif im.ndim == 3:
+        out = np.zeros_like(im)
+        for c in range(im.shape[2]):
+            try:
+                out[..., c] = _denoise_wavelet(im[..., c], wavelet=wavelet,
+                                                           threshold=threshold)
+            except ValueError:
+                raise size_error
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: See comment below

File: skimage/restoration/_denoise.py [hunk 5]
```diff
@@ -355,3 +355,92 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
         raise ValueError('only 2-d and 3-d images may be denoised with this '
                          'function')
     return out
+
+
+def denoise_wavelet(im, threshold=0.2, wavelet='db1', clip=False):
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: `clip` is not used anywhere in the code.

**Comment by soupault** [comment]
Hi @scottsievert !
Thanks for your work!
I've just added some recommendations on code improvement.

Regarding your questions:
*clipping* : it is mandatory to always return to correct image, so unless you have a better idea how to handle negative values, function has to always return image with all values within [0;1].

*sizing* : I guess you should investigate better into DWT. As far as I know, the most common implementation is based on /2 downsampling, so it is neccesary (unless another way isn't proposed) to pad the input image.

*tests* : you should add both basic tests (check if function returens error for >2D images, for example) and functional ones (test if result on some image from `skimage` returns the expected result). These are minimal set of tests we would like to see.

Regarding other points: you should also add example to gallery (check `doc/examples/` directory).

**Comment by scottsievert** [comment]
@soupault Thanks for answering my questions! Can we assume our input images scaled from 0 to 1?

8997800d12f3ad5cc45fc7f2de950ff05f44f018 fixes many of these issues. I still have to

- [ ] Implement tests
- [ ] Better checking of size issues.

**Comment by soupault** [comment]
@scottsievert yes, due to `scikit-image` conventions, once you call `im = img_as_float(im)`, you should consider you image in float numbers from interval [0;1] only.

**Comment by alexandrejaguar** [comment]
Hi guys, this interests me a lot! Thanks @scottsievert for that!
About one of @soupault hints:
* yes, the most used discrete wavelet algorithm uses downsampling by 2. You could use `np.pad` for padding. Here is how I solved that issue on my input images:

```python
    # mirroring parameter: lower pixel number
    if (m < n):
        par = m
    else:
        par = n

    # mirroring
    aux_aprx = np.pad(img, (par, par), 'symmetric')
```

Thanks again! Have a nice one!

**Comment by JDWarner** [comment]
I'm traveling but want to clarify a little here.

There is absolutely a place for optionally allowing values outside [0, 1] in this package. See for example the implementation of `skimage.util.random_noise`. This is similar; I recommend retaining the kwarg following those conventions.

Clipping should probably be the default, but it should be possible to disable for advanced users who want to do further work without throwing away data.

**Comment by grlee77** [comment]
A couple of comments on a few implementation details:

- If you are intending to apply a hard threshold, you need to pass `mode='hard'` to `pywt.threshold`.  The default is soft thresholding.

- It would be nice if rather than being forced to manually specify a threshold, some simple common threshold estimates from the literature (e.g. VisuShrink, BayesShrink, or SUREShrink) could be implemented. 

Also, there are substantial improvements in the upcoming release of `PyWavelets` that could be of use here (the current plan is to release  v0.4.0 on Dec 28). 

 - 2D and nD wavelet transforms are MUCH (up to two orders of magnitude) faster than in v0.3.0.
 - I don't think you will have to worry about the sizing issues anymore.  These should be automatically handled by the `wavedec`*/`waverec`* functions now.
 - There are now n-dimensional versions of the multilevel transforms:  `wavedecn`/`waverecn`,  so this wouldn't have to be restricted to 2D/2D+color images.

**Comment by soupault** [comment]
@grlee77 thank you very much for the comment! Version 0.4.0 sounds like a quite major update.
I guess it is better for us to target 0.4.0 so as the PR, in any case, requires some work to be done.

**Comment by scottsievert** [comment]
@grlee77 BayesShrink implemented as explained in [(1)] -- thanks for the reference. I have also removed the keyword argument `enforce_size` (and `clip` too).

[(1)]:http://www.egr.msu.edu/~aviyente/bayesshrink.pdf

File: skimage/restoration/_denoise.py [hunk 6]
```diff
@@ -355,3 +355,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+    Examples
+    --------
+    >>> from skimage import color, data
+    >>> img = data.astronaut() * 1.0 / 255
+    >>> img = color.rgb2gray(img)
+    >>> img += 0.5 * img.std() * np.random.randn(*img.shape)
+    >>> img = np.clip(img, 0, 1)
+    >>> denoised_img = denoise_wavelet(img)
+    """
+    def _denoise_wavelet(img, wavelet, threshold):
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: I believe you should move function definition outside so to avoid overhead in batch processing.

**Comment by emmanuelle** [comment]
Hi @scottsievert thanks for this useful addition! Could you please write an example for the gallery using wavelet denoising? 

Also, there is some "staircasing" in the example that you showed, which is a bit surprising. I could have expected staircasing with Haar wavelet but I reckon that have used Daubechies wavelets. Is there a way to make it look better in the example?

**Comment by alexandrejaguar** [comment]
@scottsievert @emmanuelle I think you could show two examples on that,
maybe with a lower and a higher order Daubechies wavelet. What do you think
about it?

**Comment by scottsievert** [comment]
@emmanuelle The BayesShrink helps with the "stair stepping." In the new implementation, you can (optionally) specify a noise standard deviation. Here's the output (after slightly tuning the noise standard deviation, not the threshold)

figure_1

File: skimage/restoration/_denoise.py [hunk 7]
```diff
@@ -355,3 +355,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+
+    if not im.dtype.kind == 'f':
+        im = img_as_float(im)
+
+    if threshold is None:
+        # The BayesShrink threshold from [1]_ in docstring.
+        threshold = noise_stdev**2 / np.sqrt(max(im.var() - noise_stdev**2, 0))
+
+    if im.ndim == 2:
+        out = _denoise_wavelet(img, wavelet=wavelet, threshold=threshold)
```
Comment (grlee77) [inline skimage/restoration/_denoise.py]: typo here:  `img` here should be `im`

File: skimage/restoration/_denoise.py [hunk 8]
```diff
@@ -355,3 +355,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+    .. math:: \widehat{x} = \arg \min_x ||z - x||_2^2 + \lambda ||x||_1
+
+    where :math:`z` is the input image wavelet coefficients and :math:`\lambda`
+    is the threshold.
+
+    This function performs wavelet denoising on each color plane separately.
+
+    References
+    ----------
+    .. [wiki] https://en.wikipedia.org/wiki/Wavelet#Wavelet_denoising.
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: Please enumerate these references.

**Comment by soupault** [comment]
@scottsievert hi!
Regarding the point in the header:

    generalize to n dimensional wavelet transforms (not doing it on each color plane)

I believe n-dimensionality is considered as generalization in the sense of 1D (timeseries), 2D (images) and 3D (videos) cases. Color channels form a "special" dimension and usually should be processed separately.

**Comment by JDWarner** [comment]
@soupault Not exactly. You're correct that color channels are a special case, but let's keep time out of this. For reference, the progression for n-dimensional data is best understood from 0-3 spatial dimensions where

0: Single point
1: Line
2: Plane
3: Volume

Note that n-dimensional analysis *almost never* mixes units. As we increased the dimensionality, every added dimension was orthogonal to the prior dimensions and **and** also carried the same spatial units. When you throw time into the mix things get treated very differently (time series analysis, etc.).

**Comment by scottsievert** [comment]
@JDWarner @soupault So would this be just doing this for each dimension separately (using `wavedec2`? I envisioned taking an n-dimensional wavelet transform with `wavedecn`.

**Comment by JDWarner** [comment]
@scottsievert The best API is to assume the input is n-dimensional with all spatial dimensions (i.e., use `wavedecn`) but offer a binary `multichannel=False` kwarg which people may optionally set. See `skimage.segmentation.random_walker`.

**Comment by soupault** [comment]
@JDWarner of course, you are right. I've expressed myself wrong.

At the same time, I can't imagine if someone needs to run algorithms developed especially for images on 1D (line) or 3D (volume). The only cases I can think of are 1D as timeseries (0D data, 1D time), 3D as multichanneled image (2D data, 1D channel), 3D as a videostream (2D data, 1D time) and 3D a batch processing of images (2D data, 1D non-consistent data). None of these cases actually require `nD` support, because time dimension is usually being considered and explored by researchers separately (like the channel dimension is).
I'm just trying to say that with the specification of image `multichannel` it seems far more logical to add `timechannel`, rather then prefer `nD` support. 

P.S. (dream of the better future) - With the full `timechannel` support `skimage` will be the only package everybody ever needs.

**Comment by JDWarner** [comment]
A great deal of image analysis in science and medicine concerns 3-D volumetric data, from confocal microscopy to medical imaging. And, while some may have a great deal of time series data, it's still more common to have channels (color) than video.

**Comment by JDWarner** [comment]
We've veered off topic - my fault - but let's keep things related to this PR.

**Comment by soupault** [comment]
@scottsievert Hi! `PyWavelets` 0.4.0 is finally out! Are you planning to finalize the RP?

**Comment by scottsievert** [comment]
There are other open PRs (#1837, #1838) that this PR depends on. How should I handle this?

**Comment by soupault** [comment]
@scottsievert I'd suggest to focus on the feature itself. You could finalize it (e.g. considering proposals in https://github.com/scikit-image/scikit-image/pull/1833#issuecomment-166159437) and prepare some tests, running them locally.
Once you'll be confident about this PR and `skimage` 0.12 will be released, we'll hop into https://github.com/scikit-image/scikit-image/issues/1838 and perform required modifications to the package infrastructure (prepare wheels for travis, handle optional_req, etc).

**Comment by scottsievert** [comment]
Okay, tests and examples added.

**Comment by soupault** [comment]
@stsievert thanks a lot! I think we should plan this and other wavelets-related issues/PRs for `skimage==0.13`.
BTW, I've found this old PR on the same topic: https://github.com/scikit-image/scikit-image/pull/760/ . Scott, could you, please, take a look and hopefully transfer several enhancements from it?

**Comment by stsievert** [comment]
That PR for soft or hard thresholding, and has an array of thresholds for each wavelet level. I've implemented soft/hard thresholding but am ignoring a list of different coefficients. In the problem formulations I've seen, when it's known that the wavelet coefficients have at most k non-zeros, it's optimal to choose a constant threshold.

File: skimage/restoration/_denoise.py [hunk 9]
```diff
@@ -355,3 +357,139 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
...
+        Input data to be denoised. `im` can be of any numeric type,
+        but it is cast into an ndarray of floats for the computation
+        of the denoised image.
+    wavelet : string
+        The type of wavelet to perform. Can be any of the options
+        [pywt.wavelist]_ outputs. For example, this may be any of ``{db1, db2,
+        db3, db4, haar}``.
+    noise_std : float, optional
+        The noise is estimated when noise_std is None (the default). It does
+        this as mentioned in pull request #1837.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Do we want to keep a reference to a PR in a docstring (even for a hiden function)? I think the information can be summarized in a Notes section.

**Comment by soupault** [comment]
Closed in advance of #2190.

**Pull request closed without merging**